fvm/evm/stdlib/contract.go

```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+	gauge common.MemoryGauge,
+) *interpreter.HostFunctionValue {
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		internalEVMTypeEncodeABIFunctionType,
+		func(invocation interpreter.Invocation) interpreter.Value {
+			inter := invocation.Interpreter
+			locationRange := invocation.LocationRange
+
+			// Get `values` argument
+
+			valuesArray, ok := invocation.Arguments[0].(*interpreter.ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			values := make([]interface{}, 0)
+			var arguments gethABI.Arguments
+
+			for i := 0; i < valuesArray.Count(); i++ {
+				element := valuesArray.Get(inter, locationRange, i)
+				switch value := element.(type) {
+				case *interpreter.StringValue:
+					values = append(values, value.Str)
+					typ, err := gethABI.NewType("string", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.UInt64Value:
+					values = append(values, uint64(value))
+					typ, err := gethABI.NewType("uint64", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.BoolValue:
+					values = append(values, bool(value))
+					typ, err := gethABI.NewType("bool", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case *interpreter.CompositeValue:
+					if value.QualifiedIdentifier == "EVM.EVMAddress" {
+						bytes, err := interpreter.ByteArrayValueToByteSlice(
+							inter,
+							value.GetMember(inter, locationRange, "bytes"),
+							locationRange,
+						)
+						if err != nil {
+							panic(err)
+						}
+						values = append(values, gethCommon.Address(bytes))
+						typ, err := gethABI.NewType("address", "", nil)
+						if err != nil {
+							panic(err)
+						}
+						arguments = append(arguments, gethABI.Argument{Type: typ})
+					}
+				case *interpreter.ArrayValue:
+					elements := make([]uint64, 0)
+					for i := 0; i < value.Count(); i++ {
+						element := value.Get(inter, locationRange, i)
+						v := element.(interpreter.UInt64Value)
+						elements = append(elements, uint64(v))
+					}
+					values = append(values, elements)
+					typ, err := gethABI.NewType("uint64[]", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				}
+			}
+
+			encoded, err := arguments.Pack(values...)
+			if err != nil {
+				panic(err)
+			}
+
+			return interpreter.ByteSliceToByteArrayValue(inter, encoded)
+		},
+	)
+}
+
+// EVM.decodeABI
+
+const internalEVMTypeDecodeABIFunctionName = "decodeABI"
+
+var internalEVMTypeDecodeABIFunctionType = &sema.FunctionType{
+	Parameters: []sema.Parameter{
+		{
+			Label:      sema.ArgumentLabelNotRequired,
+			Identifier: "types",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				sema.NewVariableSizedType(nil, sema.AnyType),
```

Comment:
I was not sure what `TypeAnnotation` to use for the `types` argument, given the current definition:

```cadence
access(all)
fun decodeABI(_ types: [Type], data: [UInt8]): [AnyStruct] {
    return InternalEVM.decodeABI(types, data: data)
}
```
I am probably not using the correct one :thinking:

Reply:
Updated in https://github.com/onflow/flow-go/pull/5024/commits/22ab4ae59ab0bf715cbf4751d136c6c92c782d3d